category/applicative.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Instances for identity and composition functors
+-/
+
+import category.functor
+
+universe variables u v w u' v' w'
+
+section lemmas
+
+open function applicative
+
+variables {α β γ : Type u}
+variables {f : Type u → Type v}
+variables [applicative f] [is_lawful_applicative f]
+variables (g : β → γ)
+
+lemma pure_seq_pure {α β : Type u} (g : α → β) (x : α) :
+  pure g <*> (pure x : f α) = pure (g x) :=
+by simp
+
+open function
+
+lemma applicative.map_seq {β γ σ : Type u} (h : γ → σ) (x : f (β → γ)) (y : f β) :
+  h <$> (x <*> y) = (comp h <$> x) <*> y :=
+by rw [← pure_seq_eq_map,← pure_seq_eq_map,
+       seq_assoc,map_pure]
+
+lemma applicative.seq_map {β γ σ : Type u} (h : σ → β) (x : f (β → γ)) (y : f σ) :
+  x <*> (h <$> y) = (flip comp h) <$> x <*> y :=
+by rw [← pure_seq_eq_map,← pure_seq_eq_map,seq_assoc,seq_pure,pure_seq_eq_map,← comp_map]; refl
+
+open applicative functor
+
+attribute [norm] seq_assoc pure_seq_eq_map map_pure seq_map map_seq
```

Comment:
What is this attribute? Mathlib doesn't have many simp sets defined beyond the default one, and it's not clear to me this is how we want to organize these.

Reply:
This attribute is defined in core: https://github.com/leanprover/lean/blob/master/library/init/meta/simp_tactic.lean#L506

`changes.md` says:

> Add [norm] simp set. It contains all lemmas tagged with [simp] plus any lemma tagged with [norm]. These rules are used to produce normal forms and/or reduce the number of constants used in a goal. For example, we plan to add the lemma `f <$> x = x >>= pure ∘ f` to [norm].

It simplifies proofs with applicative functors greatly.

Reply:
I added `functor_norm`. I'm not sure how helpful `norm` is in general.

Reply:
I saw your `functor_norm` mode after I chose to get into `norm`. Why is it distinct from `norm`?